### ISSUE TITLE
React: Hooks for customization instead of render props

### ISSUE THREAD

**Issue by bennettdams** [tailwindlabs/headlessui#205]
Here's a proposal: Provide hooks together with the components. It would be nice to have something like `useMenu`, `useListbox` or `useSwitch` in the components where Headless UI components are used.

Basic example:
I want to use the `Menu` component in my navbar/header, so I want to show the dropdown area also when I'm hovering the button (instead of clicking). You could add hover handlers and refs to check the hover state yourself, but wouldn't it be nice to receive a hook out of the box? Basic examples:
```tsx
const MyComponent = () => {
  const { isChecked, toggle } = useSwitch();
  const { isOpen, onHover } = useMenu();

  return (
    <div>
      <Menu>
        <Menu.Button>More</Menu.Button>
        ...
      </Menu>
    </div>
  );
};
```
These simple customizations (e.g. "hover state") are _easy_ to implement for the developers of Headless UI, but hard to do for developers who are using Headless UI. They would allow way more flexible use cases for the components.

**Comment by RobinMalfait** [comment]
Hey! Thank you for your suggestion!
Much appreciated! 🙏

We've considered hooks in our API design, and maybe we want to expose hooks in the future, but not for now.
I love hooks, and we also use them internally, but as a public API for a component library they have some flaws, and here are some of the reasons why we didn't use hooks:

### They can force you to abstract components

For example, when you have a table with users, and each user has a `Menu` component with some actions. This is all good, and your code could look like  this:

```jsx
function MyComponent() {
  let users = []

  return (
    <table>
      <tbody>
        {users.map(user => (
          <tr key={user.id}>
            <td>{user.name}</td>
            <td>
              <Menu />
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  )
}
``` 
If we used hooks, then the data in the `.map()` would have been it's own component, because you can't use hooks conditionally and thus also not inside a loop.

```jsx
function MyDropdown() {
  let { open } = useMenu()
  return <Menu />
}

function MyComponent() {
  let users = []

  return (
    <table>
      <tbody>
        {users.map(user => (
          <tr key={user.id}>
            <td>{user.name}</td>
            <td>
              <MyDropdown />
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  )
}
```

While this is okay, this forces you to abstract components and think of a name for that component. This is fine and makes sense for some use cases, not for others. You would also need to figure out some mechanism to get the `open` state back into the main `MyComponent` if you wanted to use it there.

### There is no link between a hook and a component

In the example you showed, there is no immediate "link" between the component and the hook itself. This means that you can't use multiple instances of this component.

```jsx
function MyComponent() {
  // To which menu is this "state" linked?
  let { open, active } = useMenu()

  return (
    <div>
      <Menu>Menu 1</Menu>
      <Menu>Menu 2</Menu>
    </div>
  )
}
```

This can of course be fixed in a lot of different ways, here are a few potential fixes:

1. You can use multiple hooks, and expose its data and a component from the hook

    ```jsx
    function MyComponent() {
      let { Menu: Menu1, open: open1, active: active1 } = useMenu()
      let { Menu: Menu2, open: open2, active: active2 } = useMenu()
    
      return (
        <div>
          <Menu1>Menu 1</Menu1>
          <Menu2>Menu 2</Menu2>
        </div>
      )
    }
    ```

    This is not very elegant, but it could work.
2. You can spread props onto your component:

    ```jsx
    function MyComponent() {
      let { open: open1, active: active1, menuProps: menuProps1 } = useMenu()
      let { open: open2, active: active2, menuProps: menuProps2 } = useMenu()
    
      return (
        <div>
          <Menu {...menuProps1}>Menu 1</Menu>
          <Menu {...menuProps2}>Menu 2</Menu>
        </div>
      )
    }
    ```

    A lot of component libraries do something similar. This works, but it has the same issues that you have to start naming things. There is also a possibility that you "forget" to spread the props onto your component. 

These examples also don't work when you are using them in a loop. Once you use it in a loop, you probably have to abstract components again.

### Full "page" re-renders

I know that this is not an issue if you abstract your components, but if you don't then you can re-render the whole page. For example, when you have a `Menu` component, when you hover or use your arrow keys to highlight a `MenuItem` then we expose some state. This state internally triggered a re-render which is fine.

In our current implementation this means that the `Menu` component will re-render to update all its children. But imagine that you now have a big landing pages or a marketing page or just a page that contains a lot of stuff. Once you use hooks, that whole component will re-render, not only the `Menu`.

I know that this might not be a problem (because your re-renders should be cheap anyway), and that it is definitely not a problem when you abstract components, but there are scenarios where this _is_ a problem. 

### API Consistency

I know that this is not really an issue for you as a React user. But for us it is nice to have a very similar API between React and Vue components. This makes it easier to create examples, talk about it, or just provide help in general.

I know that this is not a 100% valid argument, because if React or Vue exposes an API that's perfect for this situation then we would be stupid to not make use of this. But since we have a bunch of "issues" / "concerns" currently, we opted for plain old components.

---

Long story short, one day we might explore how to expose hooks or similar mechanisms. But for now they will stay as components.
